lib/babel/transformation/transformers/spec/function-name.js

```diff
@@ -0,0 +1,53 @@
+"use strict";
+
+var t           = require("../../../types");
+var util        = require("../../../util");
+
+var propertyFunctionVisitor = {
+  enter: function (node, parent, scope, state, file) {
+    if (t.isReferencedIdentifier(node, parent, { name: state.name }) && scope.getBindingIdentifier(node.name) === state.binding) {
+      return state.getOuter();
+    }
+  }
+};
+
+exports.FunctionExpression = function (node, parent, scope, file) {
+  if (node.id) return;
+  var id;
+  if (t.isProperty(parent)) {
+    id = parent.key;
+  } else if (t.isVariableDeclarator(parent)) {
+    id = parent.id;
+  } else {
+    return;
+  }
+  var binding = scope.getBindingIdentifier(id.name);
+  var outerId, selfGlobalId;
+  scope.traverse(node, propertyFunctionVisitor, {
+    name: id.name,
+    binding: binding,
+
+    getOuter: function () {
+      if (!binding) {
+        return t.memberExpression(
+          selfGlobalId || (selfGlobalId = file.addHelper("self-global")),

```

Comment:
Yeah but just leaving in the reference instead of accessing the global
object will surely work?

On Wednesday, 18 February 2015, Ingvar Stepanyan <notifications@github.com>
wrote:

> In lib/babel/transformation/transformers/spec/function-name.js
> <https://github.com/babel/babel/pull/785#discussion_r24883194>:
>
> > +    id = parent.key;
> > +  } else if (t.isVariableDeclarator(parent)) {
> > +    id = parent.id;
> > +  } else {
> > +    return;
> > +  }
> > +  var binding = scope.getBindingIdentifier(id.name);
> > +  var outerId, selfGlobalId;
> > +  scope.traverse(node, propertyFunctionVisitor, {
> > +    name: id.name,
> > +    binding: binding,
> > +
> > +    getOuter: function () {
> > +      if (!binding) {
> > +        return t.memberExpression(
> > +          selfGlobalId || (selfGlobalId = file.addHelper("self-global")),
>
> @sebmck <https://github.com/sebmck> Because variable can be defined as
> global in other scripts and so exist as property on window/global, and
> Babel is unaware of those.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/pull/785/files#r24883194>.
>


-- 
Sebastian McKenzie